### PR TITLE
feat(decode): add tolerantDecoding option

### LINKED ISSUE #56: DecodeBlock: Cannot read property '0' of undefined
Hello,

I'm using jimp, which use your library. I have a problem with some images, which, when I want to preview them with jimp, throw a JS error:

```
TypeError: Cannot read property '0' of undefined
    at decodeBlock (XXXXXXXXXXX\node_modules\jpeg-js\lib\decoder.js:264:51)
    at decodeScan (XXXXXXXXXXX\node_modules\jpeg-js\lib\decoder.js:298:11)
    at constructor.parse (XXXXXXXXXXX\node_modules\jpeg-js\lib\decoder.js:744:29)
    at Object.decode [as image/jpeg] (XXXXXXXXXXX\node_modules\jpeg-js\lib\decoder.js:1004:11)
    at Jimp.parseBitmap (XXXXXXXXXXX\node_modules\@jimp\core\dist\utils\image-bitmap.js:115:53)
    at Jimp.parseBitmap (XXXXXXXXXXX\node_modules\@jimp\core\dist\index.js:521:32)
    at new Jimp (XXXXXXXXXXX\node_modules\@jimp\core\dist\index.js:474:13)
    at _construct (XXXXXXXXXXX\node_modules\@babel\runtime\helpers\construct.js:30:21)
    at XXXXXXXXXXX\node_modules\@jimp\core\dist\index.js:1016:32
    at new Promise (<anonymous>)
    at Function.Jimp.read (XXXXXXXXXXX\node_modules\@jimp\core\dist\index.js:1015:10)
    ...
```

I'm pretty sure this is due to the fact that the image must have an invalid format, but it is nevertheless read by all the readers I've tested (Chrome, Paint, Photoshop, Windows Photos, ...).

And after trying to update your lib, so as not to call the function if the value is undefined, it works perfectly (my preview is generated by jimp):

```js
function decodeBlock(component, decode, mcu) {
  var blockRow = (mcu / component.blocksPerLine) | 0;
  var blockCol = mcu % component.blocksPerLine;
  if (component.blocks[blockRow] !== undefined) {
    decode(component, component.blocks[blockRow][blockCol]);
  }
}
```

I'm doing this issue rather than a PR because I don't know anything about the JPEG format, and I don't know if it can have unintended consequences. What do you think about that? Is it possible to fix it? If my `if` is enough, I can of course do the PR to help you.

_I would have appreciated being able to share these images to show you an example of the issue, but these images are private. And if I edit them to remove the private parts, the problem disappears..._

### PR DESCRIPTION
**Pull request by Dewep**
closes #56

### CODE REVIEW

File: lib/decoder.js [hunk 1]
```diff
@@ -256,12 +256,16 @@ var JpegImage = (function jpegImage() {
       var mcuCol = mcu % mcusPerLine;
       var blockRow = mcuRow * component.v + row;
       var blockCol = mcuCol * component.h + col;
-      decode(component, component.blocks[blockRow][blockCol]);
+      if (component.blocks[blockRow] !== undefined || opts.tolerantDecoding !== true) {
```
Comment (patrickhulce) [inline lib/decoder.js]:
```suggestion
      // If the block is missing and we're in tolerant mode, just skip it.
      if (component.blocks[blockRow] === undefined && opts.tolerantDecoding) return;
```

WDYT?

File: lib/decoder.js [hunk 2]
```diff
@@ -256,12 +256,16 @@ var JpegImage = (function jpegImage() {
...
+        decode(component, component.blocks[blockRow][blockCol]);
+      }
     }
     function decodeBlock(component, decode, mcu) {
       var blockRow = (mcu / component.blocksPerLine) | 0;
       var blockCol = mcu % component.blocksPerLine;
-      decode(component, component.blocks[blockRow][blockCol]);
+      if (component.blocks[blockRow] !== undefined || opts.tolerantDecoding !== true) {
```
Comment (patrickhulce) [inline lib/decoder.js]:
```suggestion
      // If the block is missing and we're in tolerant mode, just skip it.
      if (component.blocks[blockRow] === undefined && opts.tolerantDecoding) return;
```

File: lib/decoder.js [hunk 3]
```diff
@@ -980,15 +984,18 @@ module.exports = decode;
 function decode(jpegData, opts) {
   var defaultOpts = {
     useTArray: false,
-    colorTransform: true
+    colorTransform: true,
+    tolerantDecoding: true
```
Comment (patrickhulce) [inline lib/decoder.js]: to avoid any future breaking change based on tolerantDecoding, how do you feel about this being off by default?

**Comment by Dewep** [comment]
I will have enabled the default option since this is the behavior of other readers, otherwise the lib will crash in all cases. But I understand your hesitation, it is not disturbing to turn it off.

I changed the default value, and made the small corrections you suggested.

I've searched the internet for images that would have the same issue, without success. And I'm not able to make this picture myself (I clearly don't know enough JPEG for that). The private images in my possession that have this concern all seem to have a name in the form `SKMBT_C22419122709450_0018.jpg`. I did find some public images with a similar name, but they all work.

**Pull request merged**